### PR TITLE
Terrajet AWS Provider for Crossplane

### PR DESCRIPTION
**Pull request by vara-bonthu**
### What does this PR do?

- `Terrajet AWS Provider` for Crossplane added
- Options to deploy `Default AWS provider` or `Terrajet AWS Provider` for Crossplane
- Docs Updated
- Examples updated

### More

- [x] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [x] Yes, I have added a new example under [examples](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/ssp-eks-add-ons) repo (if applicable) - `Not required for this PR`
- [x] Yes, I have updated the [docs](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

**Review by askulkarni2 — APPROVED** [review]

File: docs/add-ons/crossplane.md [hunk 1]
```diff
@@ -39,26 +39,40 @@ You can optionally customize the Helm chart that deploys `Crossplane` via the fo
 ```
 
 ### Crossplane AWS Provider Deployment
-AWS Provider for Crossplane gets deployed by default when you enable `enable_crossplane = true`.
-The below configuration helps you to upgrade the AWS provider version and lets you define custom IAM policies to manage AWS resources through IRSA.
+This module provides options to deploy the following AWS providers for Crossplane. These providers disabled by default, and it can be enabled using the config below.
 
-Crossplane requires Admin like permissions to create and update resources similar to Terraform deploy role.
+ - [AWS Provider](https://github.com/crossplane/provider-aws)
+ - [Provider Jet AWS](https://github.com/crossplane-contrib/provider-jet-aws)
```
Comment (kcoleman731) [inline docs/add-ons/crossplane.md:45]: Terrajet AWS Provider is the name in the repo?

File: examples/crossplane/README.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ This example deploys the following Basic EKS Cluster with VPC
  - Creates EKS Cluster Control plane with one managed node group
  - Crossplane Add-on to EKS Cluster
  - AWS Provider for Crossplane
+ - Jet AWS Provider for Crossplane
```
Comment (kcoleman731) [inline examples/crossplane/README.md:8]: Terrajet?

File: modules/kubernetes-addons/crossplane/main.tf [hunk 3]
```diff
@@ -17,49 +17,127 @@ module "helm_addon" {
...
-  yaml_body = templatefile("${path.module}/aws-provider/controller-config.yaml", {
-    iam-role-arn = "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:role/${var.eks_cluster_id}-provider-aws--irsa"
+#--------------------------------------
+# AWS Provider
+#--------------------------------------
+resource "kubectl_manifest" "aws_controller_config" {
+  count = var.aws_provider.enable == true ? 1 : 0
+
+  yaml_body = templatefile("${path.module}/aws-provider/aws-controller-config.yaml", {
+    iam-role-arn = "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:role/${var.eks_cluster_id}-${local.aws_provider_sa}-irsa"
```
Comment (kcoleman731) [inline modules/kubernetes-addons/crossplane/main.tf:27]: we have `arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:role/` in lots of places in this codebase. We should really think about putting this in a variable.

File: modules/kubernetes-addons/crossplane/main.tf [hunk 4]
```diff
@@ -17,49 +17,127 @@ module "helm_addon" {
...
   description = "Crossplane AWS Provider IAM policy"
-  name        = "${var.eks_cluster_id}-aws-provider-irsa"
+  name        = "${var.eks_cluster_id}-${local.aws_provider_sa}-irsa"
   policy      = data.aws_iam_policy_document.s3_policy.json
   tags        = var.tags
 }
 
+resource "kubectl_manifest" "aws_provider_config" {
+  count = var.aws_provider.enable == true ? 1 : 0
+
```
Comment (kcoleman731) [inline modules/kubernetes-addons/crossplane/main.tf:68]: new lines?

File: modules/kubernetes-addons/crossplane/main.tf [hunk 5]
```diff
@@ -17,49 +17,127 @@ module "helm_addon" {
...
+  yaml_body = templatefile("${path.module}/aws-provider/aws-provider-config.yaml", {})
+
+  depends_on = [kubectl_manifest.aws_provider, time_sleep.wait_30_seconds_aws]
+}
+
 # Wait for the AWS Provider CRDs to be fully created before initiating aws_provider_config deployment
-resource "time_sleep" "wait_30_seconds" {
-  depends_on = [kubectl_manifest.aws_provider]
+resource "time_sleep" "wait_30_seconds_aws" {
```
Comment (kcoleman731) [inline modules/kubernetes-addons/crossplane/main.tf:75]: What is the purpose of the sleep? A random 30 sec sleep seems like a bad idea?

File: modules/kubernetes-addons/crossplane/main.tf [hunk 6]
```diff
@@ -17,49 +17,127 @@ module "helm_addon" {
...
+#--------------------------------------
+# Terrajet AWS Provider
+#--------------------------------------
+resource "kubectl_manifest" "jet_aws_controller_config" {
+  count = var.jet_aws_provider.enable == true ? 1 : 0
 
-resource "kubectl_manifest" "aws_provider_config" {
-  yaml_body = templatefile("${path.module}/aws-provider/aws-provider-config.yaml", {})
+  yaml_body = templatefile("${path.module}/aws-provider/jet-aws-controller-config.yaml", {
+    iam-role-arn = "arn:${data.aws_partition.current.partition}:iam::${data.aws_caller_identity.current.account_id}:role/${var.eks_cluster_id}-${local.jet_aws_provider_sa}-irsa"
```
Comment (kcoleman731) [inline modules/kubernetes-addons/crossplane/main.tf:88]: see comment above on this value

File: modules/kubernetes-addons/crossplane/main.tf [hunk 7]
```diff
@@ -17,49 +17,127 @@ module "helm_addon" {
...
+resource "kubectl_manifest" "jet_aws_provider_config" {
+  count = var.jet_aws_provider.enable == true ? 1 : 0
+
+  yaml_body = templatefile("${path.module}/aws-provider/jet-aws-provider-config.yaml", {})
+
+  depends_on = [kubectl_manifest.jet_aws_provider, time_sleep.wait_30_seconds_jet_aws]
+}
+
+# Wait for the AWS Provider CRDs to be fully created before initiating jet_aws_provider_config deployment
+resource "time_sleep" "wait_30_seconds_jet_aws" {
```
Comment (kcoleman731) [inline modules/kubernetes-addons/crossplane/main.tf:138]: same comment as above

File: modules/kubernetes-addons/crossplane/main.tf [hunk 8]
```diff
@@ -17,49 +17,127 @@ module "helm_addon" {
...
+  description = "Crossplane Jet AWS Provider IAM policy"
+  name        = "${var.eks_cluster_id}-${local.jet_aws_provider_sa}-irsa"
+  policy      = data.aws_iam_policy_document.s3_policy.json
+  tags        = var.tags
+}
+
+resource "kubectl_manifest" "jet_aws_provider_config" {
+  count = var.jet_aws_provider.enable == true ? 1 : 0
+
+  yaml_body = templatefile("${path.module}/aws-provider/jet-aws-provider-config.yaml", {})
```
Comment (kcoleman731) [inline modules/kubernetes-addons/crossplane/main.tf:132]: new lines?

File: docs/add-ons/crossplane.md (hunk 1 above)
Comment (vara-bonthu) [reply docs/add-ons/crossplane.md:45]: It's not the name of the repo but they mentioned this(`Terrajet AWS Provider`) in README. I will refer this everywhere in the docs

File: modules/kubernetes-addons/crossplane/main.tf (hunk 3 above)
Comment (vara-bonthu) [reply modules/kubernetes-addons/crossplane/main.tf:27]: I will use local variable to replace this. This way we only make one call to the data resource API. This is my assumption but we can check with hashicorp as well.

File: modules/kubernetes-addons/crossplane/main.tf (hunk 5 above)
Comment (vara-bonthu) [reply modules/kubernetes-addons/crossplane/main.tf:75]: I have added a comment above with the justification. kubectl manifest applies the manifest in seconds and comes back with the response as complete however the CRDs are still being created behind the scenes.

This is one of the race condition which i am handling using this time_sleep. This normally takes 10 to 20 seconds to create all the CRDs which is a dependency to the next resource(provider config). The next resource apply fails if these are not ready.  

We can improve this if we find a better solution later .

File: examples/crossplane/main.tf [hunk 9]
```diff
@@ -143,9 +143,17 @@ module "kubernetes-addons" {
   # Refer to docs/add-ons/crossplane.md for advanced configuration
   enable_crossplane = true
 
-  # Optional config to deploy specific version of AWS Provider and attach additional IAM policies to manage AWS resources using Crossplane
-  crossplane_provider_aws = {
-    provider_aws_version     = "v0.23.0"
+  # Creates ProviderConfig -> aws-provider
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply examples/crossplane/main.tf:147]: done

File: docs/add-ons/crossplane.md (hunk 1 above)
Comment (vara-bonthu) [reply docs/add-ons/crossplane.md]: done

File: modules/kubernetes-addons/crossplane/main.tf (hunk 3 above)
Comment (vara-bonthu) [reply modules/kubernetes-addons/crossplane/main.tf]: done

File: modules/kubernetes-addons/crossplane/main.tf (hunk 6 above)
Comment (vara-bonthu) [reply modules/kubernetes-addons/crossplane/main.tf]: done

File: modules/kubernetes-addons/crossplane/main.tf (hunk 8 above)
Comment (vara-bonthu) [reply modules/kubernetes-addons/crossplane/main.tf:124]: done

File: modules/kubernetes-addons/crossplane/main.tf (hunk 4 above)
Comment (vara-bonthu) [reply modules/kubernetes-addons/crossplane/main.tf]: done

File: examples/crossplane/README.md (hunk 2 above)
Comment (vara-bonthu) [reply examples/crossplane/README.md]: updated

File: modules/kubernetes-addons/crossplane/main.tf (hunk 5 above)
Comment (Zvikan) [reply modules/kubernetes-addons/crossplane/main.tf:70]: We should follow https://github.com/gavinbunney/terraform-provider-kubectl/issues/41

**Review by Zvikan — APPROVED** [review]

**Pull request merged**